File: backend/emailgen/data_loader.py
"""用户/受众记录数据模型

说明：原始 email-automation 的 data_loader 依赖 JSONL 文件 + 很多 CRM 字段（gender/age_range/device...）。
集成进 CartBack 后，邮件草稿由 IGDE Agent 方案卡产出（subject/body/discount/audience/brand 等），
受众画像信息可能不全，这里把所有字段都做成可空默认值，并新增 `from_plan_card()` 工厂函数，
让 Node 传入的 plan card 可以不经 JSONL 直接送入后续文案/图片管线。
"""
from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Dict, Iterator, Optional


@dataclass
class UserRecord:
    """单收件人记录（一条受众画像 + 营销参数）"""
    user_id: str = "anon"
    email: str = ""
    brand: str = "CartBack"
    gender: str = "O"            # M / F / O
    age_range: str = "25-34"
    device: str = "iPhone"
    product: str = ""
    product_cn: str = ""
    product_en: str = ""
    discount: float = 8.0
    goal: str = "abandonment_recovery"
    send_window: str = "10:30-21:00"
    locale: str = "en-US"
    preferred_language: str = ""   # 用户习惯语言（如 "German"/"French"/"Spanish"）；空=按 locale 推断，非空=权威覆盖 locale 语言猜测
    price_sensitivity: str = ""    # 价格敏感度: value / standard / premium（标签池预留；from_plan_card 暂不采集 → 初次用户设置不显示）
    customer_segment: str = ""     # 客户分层: new / returning / vip（标签池预留；from_plan_card 暂不采集 → 初次用户设置不显示）
    cart_url: str = "https://cartback.demo"
    raw: Dict[str, Any] = field(default_factory=dict)

    @classmethod
    def from_plan_card(cls, card: Dict[str, Any], draft: Optional[Dict[str, Any]] = None) -> "UserRecord":
        """从 Node 传入的方案卡 + draft 构造用户记录

        card 字段约定（兼容 CartBack IGDE 输出）:
          subject, body, discount, audience, brand, locale, coupon, posters, cta
        """
        discount_raw = card.get("discount") or (draft or {}).get("discount") or 8
        try:
            discount = float(discount_raw)
        except (TypeError, ValueError):
            discount = 8.0
        audience = str(card.get("audience") or "").strip()
        locale = str(card.get("locale") or (draft or {}).get("locale") or "en").strip()
        if len(locale) == 2:
            locale = f"{locale}-{locale.upper()}"
        preferred_language = str(card.get("preferred_language") or (draft or {}).get("preferred_language") or "").strip()
        # 从 audience 描述里猜一个产品词给 image prompt 用（兜底）
        product_en = str(card.get("product") or card.get("product_en") or "Premium Phone Case").strip()
        product_cn = str(card.get("product_cn") or "").strip()
        brand = str(card.get("brand") or (draft or {}).get("brand") or "CartBack").strip() or "CartBack"
        cart_url = str(card.get("cart_url") or (draft or {}).get("cart_url") or "https://cartback.demo").strip()
        uid = str((draft or {}).get("id") or card.get("id") or f"dr_{abs(hash(audience + brand)) % 1_000_000:06d}")
        return cls(
            user_id=uid,
            email="",  # 草稿阶段还不知道具体收件人，真实发送时按 audience 展开
            brand=brand,
            product_en=product_en,
            product_cn=product_cn,
            product=product_en,
            discount=discount,
            locale=locale,
            preferred_language=preferred_language,
            cart_url=cart_url,
            raw={**dict(card), "draft": draft or {}},
        )


def load_user_data(file_path: str) -> Iterator[UserRecord]:
    """逐行加载 JSONL（保留原版能力，便于批量生成场景）"""
    path = Path(file_path)
    if not path.exists():
        raise FileNotFoundError(f"用户数据文件不存在: {file_path}")

    with open(path, "r", encoding="utf-8") as f:
        for line_num, line in enumerate(f, 1):
            line = line.strip()
            if not line:
                continue
            try:
                data = json.loads(line)
                yield UserRecord(
                    user_id=data.get("user_id", f"unknown_{line_num}"),
                    email=data.get("email", ""),
                    brand=data.get("brand", "CartBack"),
                    gender=data.get("gender", "O"),
                    age_range=data.get("age_range", "25-34"),
                    device=data.get("device", "iPhone"),
                    product=data.get("product", ""),
                    product_cn=data.get("product_cn", ""),
                    product_en=data.get("product_en", data.get("product", "")),
                    discount=float(data.get("discount", 0)),
                    goal=data.get("goal", "abandonment_recovery"),
                    send_window=data.get("send_window", "10:30-21:00"),
                    locale=data.get("locale", "en-US"),
                    preferred_language=data.get("preferred_language", ""),
                    price_sensitivity=data.get("price_sensitivity", ""),
                    customer_segment=data.get("customer_segment", ""),
                    cart_url=data.get("cart_url", "https://cartback.demo"),
                    raw=data,
                )
            except json.JSONDecodeError as e:
                print(f"[警告] 第 {line_num} 行 JSON 解析失败: {e}")


def count_users(file_path: str) -> int:
    count = 0
    with open(file_path, "r", encoding="utf-8") as f:
        for line in f:
            if line.strip():
                count += 1
    return count

File: backend/ts/data-loader.ts
/**
 * 用户/受众记录数据模型
 *
 * 集成进 CartBack 后，邮件草稿由 IGDE Agent 方案卡产出（subject/body/discount/audience/brand 等），
 * 受众画像信息可能不全，这里把所有字段都做成可空默认值，并新增 from_plan_card() 工厂函数。
 * 端口自 Python emailgen/data_loader.py。
 */
import * as fs from 'fs';
import * as readline from 'readline';

export interface UserRecord {
  user_id: string;
  email: string;
  brand: string;
  gender: string; // M / F / O
  age_range: string;
  device: string;
  product: string;
  product_cn: string;
  product_en: string;
  discount: number;
  goal: string;
  send_window: string;
  locale: string;
  preferred_language: string;
  price_sensitivity: string;
  customer_segment: string;
  style_preference: string; // tech / fashion / business / outdoor（受众风格品类标签代表值）
  cart_url: string;
  raw: Record<string, unknown>;
}

function strVal(v: unknown, def = ''): string {
  if (v === undefined || v === null) return def;
  return String(v);
}

/** 稳定字符串哈希（Python 侧用 hash()，进程随机；此处用确定性哈希做兜底 uid） */
function stableHash(s: string): number {
  let h = 0;
  for (let i = 0; i < s.length; i++) {
    h = (Math.imul(31, h) + s.charCodeAt(i)) | 0;
  }
  return Math.abs(h);
}

/** 标签分布快照取某类 count 最高的代表值（tagDistribution 已按 count 降序，同 type 首条即最高）。 */
function topTagByType(dist: unknown, type: string): string {
  if (!Array.isArray(dist)) return '';
  for (const item of dist) {
    if (item && typeof item === 'object' && String((item as Record<string, unknown>).tag_type || '') === type) {
      return String((item as Record<string, unknown>).tag_value || '').trim();
    }
  }
  return '';
}

export function fromPlanCard(card: Record<string, unknown>, draft?: Record<string, unknown> | null): UserRecord {
  const d = draft || {};
  const discountRaw = (card.discount ?? d.discount ?? 8) as number | string;
  let discount = 8;
  const parsed = Number(discountRaw);
  if (!Number.isNaN(parsed)) discount = parsed; // 非数字 → 维持默认 8（对应 Python 的 except → 8.0）

  const audience = strVal(card.audience).trim();
  let locale = strVal(card.locale ?? d.locale, 'en').trim();
  if (locale.length === 2) locale = `${locale}-${locale.toUpperCase()}`;
  const preferredLanguage = strVal(card.preferred_language ?? d.preferred_language).trim();
  const productEn = strVal(card.product ?? card.product_en, 'Premium Phone Case').trim();
  const productCn = strVal(card.product_cn).trim();
  const brandRaw = strVal(card.brand ?? d.brand, 'CartBack').trim();
  const brand = brandRaw || 'CartBack';
  const cartUrl = strVal(card.cart_url ?? d.cart_url, 'https://cartback.demo').trim();
  const uid = strVal(d.id ?? card.id) || `dr_${String(stableHash(audience + brand) % 1_000_000).padStart(6, '0')}`;

  // 受众标签分布快照（server.js 传入）：每类取 count 最高代表值填充画像——
  // 此前恒为硬编码默认值（O/25-34/iPhone/空），新 5 维标签实际没进文案与图片 prompt
  const dist = card.tag_distribution ?? d.tag_distribution;
  const tagGender = topTagByType(dist, 'gender');
  const tagAge = topTagByType(dist, 'age_range');
  const tagDevice = topTagByType(dist, 'device');

  return {
    user_id: uid,
    email: '',
    brand,
    product_en: productEn,
    product_cn: productCn,
    product: productEn,
    discount,
    locale,
    preferred_language: preferredLanguage,
    cart_url: cartUrl,
    gender: tagGender === 'male' ? 'M' : tagGender === 'female' ? 'F' : 'O',
    age_range: tagAge || '25-34',
    device: tagDevice || 'iPhone',
    goal: 'abandonment_recovery',
    send_window: '10:30-21:00',
    price_sensitivity: topTagByType(dist, 'price_sensitivity'),
    customer_segment: topTagByType(dist, 'customer_segment'),
    style_preference: topTagByType(dist, 'style_preference'),
    raw: { ...card, draft: d },
  };
}

/** 构造一个可完全自定义字段的 UserRecord（selftest5 用，绕过 from_plan_card） */
export function makeUser(overrides: Record<string, unknown>): UserRecord {
  const base: UserRecord = {
    user_id: 'anon',
    email: '',
    brand: 'CartBack',
    gender: 'O',
    age_range: '25-34',
    device: 'iPhone',
    product: '',
    product_cn: '',
    product_en: '',
    discount: 8,
    goal: 'abandonment_recovery',
    send_window: '10:30-21:00',
    locale: 'en-US',
    preferred_language: '',
    price_sensitivity: '',
    customer_segment: '',
    style_preference: '',
    cart_url: 'https://cartback.demo',
    raw: {},
  };
  for (const k of Object.keys(overrides)) {
    const v = overrides[k];
    if (v !== undefined && k in base) (base as unknown as Record<string, unknown>)[k] = v;
  }
  return base;
}

export interface JsonlLoadResult {
  records: UserRecord[];
  warnings: string[];
}

/** 逐行加载 JSONL（保留原版能力，便于批量生成场景） */
export async function loadUserData(filePath: string): Promise<JsonlLoadResult> {
  if (!fs.existsSync(filePath)) throw new Error(`用户数据文件不存在: ${filePath}`);
  const records: UserRecord[] = [];
  const warnings: string[] = [];
  const stream = fs.createReadStream(filePath, { encoding: 'utf8' });
  const rl = readline.createInterface({ input: stream, crlfDelay: Infinity });
  let lineNum = 0;
  for await (const line of rl) {
    lineNum++;
    const trimmed = line.trim();
    if (!trimmed) continue;
    try {
      const data = JSON.parse(trimmed);
      records.push({
        user_id: strVal(data.user_id, `unknown_${lineNum}`),
        email: strVal(data.email),
        brand: strVal(data.brand, 'CartBack'),
        gender: strVal(data.gender, 'O'),
        age_range: strVal(data.age_range, '25-34'),
        device: strVal(data.device, 'iPhone'),
        product: strVal(data.product),
        product_cn: strVal(data.product_cn),
        product_en: strVal(data.product_en ?? data.product),
        discount: Number(data.discount ?? 0) || 0,
        goal: strVal(data.goal, 'abandonment_recovery'),
        send_window: strVal(data.send_window, '10:30-21:00'),
        locale: strVal(data.locale, 'en-US'),
        preferred_language: strVal(data.preferred_language),
        style_preference: strVal(data.style_preference),
        price_sensitivity: strVal(data.price_sensitivity),
        customer_segment: strVal(data.customer_segment),
        cart_url: strVal(data.cart_url, 'https://cartback.demo'),
        raw: data,
      });
    } catch (e) {
      warnings.push(`第 ${lineNum} 行 JSON 解析失败: ${(e as Error).message}`);
    }
  }
  return { records, warnings };
}

export function countUsers(filePath: string): number {
  if (!fs.existsSync(filePath)) return 0;
  let count = 0;
  for (const line of fs.readFileSync(filePath, 'utf8').split('\n')) {
    if (line.trim()) count++;
  }
  return count;
}
